tests/components/demo/test_notify.py

```diff
@@ -8,189 +9,198 @@
+        """Record calls."""
+        calls.append(args)
+
+    return record_calls
+
+
+@pytest.fixture(name="mock_demo_notify")
+def mock_demo_notify_fixture():
+    """Mock demo notify service."""
+    with patch("homeassistant.components.demo.notify.get_service", autospec=True) as ns:
+        yield ns
+
+
+async def setup_notify(hass):
+    """Test setup."""
+    with assert_setup_component(1, notify.DOMAIN) as config:
+        assert await async_setup_component(hass, notify.DOMAIN, CONFIG)
+    assert config[notify.DOMAIN]
+    await hass.async_block_till_done()
+
+
+async def test_no_notify_service(hass, mock_demo_notify, caplog):
+    """Test missing platform notify service instance."""
+    caplog.set_level(logging.ERROR)
+    mock_demo_notify.return_value = None
+    await setup_notify(hass)
+    await hass.async_block_till_done()
+    assert mock_demo_notify.called
+    assert "Failed to initialize notification service demo" in caplog.text
+
+
+async def test_discover_notify(hass, mock_demo_notify):
+    """Test discovery of notify demo platform."""
+    assert notify.DOMAIN not in hass.config.components
+    mock_demo_notify.return_value = None
+    discovery.load_platform(
+        hass, "notify", "demo", {"test_key": "test_val"}, {"notify": {}}
+    )
+    await hass.async_block_till_done()
+    assert notify.DOMAIN in hass.config.components
+    assert mock_demo_notify.called
+    assert mock_demo_notify.mock_calls[0][1] == (
+        hass,
+        {},
+        {"test_key": "test_val"},
+    )
+
+
+async def test_sending_none_message(hass, events):
+    """Test send with None as message."""
+    await setup_notify(hass)
+    with pytest.raises(vol.Invalid):
+        await hass.services.async_call(
+            notify.DOMAIN, notify.SERVICE_NOTIFY, {notify.ATTR_MESSAGE: None}
+        )
+    await hass.async_block_till_done()
+    assert len(events) == 0
+
+
+async def test_sending_templated_message(hass, events):
+    """Send a templated message."""
+    await setup_notify(hass)
+    hass.states.async_set("sensor.temperature", 10)
+    data = {
+        notify.ATTR_MESSAGE: "{{states.sensor.temperature.state}}",
+        notify.ATTR_TITLE: "{{ states.sensor.temperature.name }}",
+    }
+    await hass.services.async_call(notify.DOMAIN, notify.SERVICE_NOTIFY, data)
+    await hass.async_block_till_done()
+    last_event = events[-1]
+    assert last_event.data[notify.ATTR_TITLE] == "temperature"
+    assert last_event.data[notify.ATTR_MESSAGE] == 10
+
+
+async def test_method_forwards_correct_data(hass, events):
+    """Test that all data from the service gets forwarded to service."""
+    await setup_notify(hass)
+    data = {
+        notify.ATTR_MESSAGE: "my message",
+        notify.ATTR_TITLE: "my title",
+        notify.ATTR_DATA: {"hello": "world"},
+    }
+    await hass.services.async_call(notify.DOMAIN, notify.SERVICE_NOTIFY, data)
+    await hass.async_block_till_done()
+    assert len(events) == 1
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6356951)